Recover peers using history from Lucene

DaveCTurner opened the pull request:
Thanks to peer recovery retention leases we now retain the history needed to
perform peer recoveries from the index instead of from the translog. This
commit adjusts the peer recovery process to do so, and also adjusts it to use
the existence of a retention lease to decide whether or not to attempt an
operations-based recovery.

Reverts #38904 and #42211
Relates #41536

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented on server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -233,20 +270,19 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
+                            // just established on the replica. This primary is certainly retaining such history, but other replicas might
+                            // not be. No big deal if this recovery succeeds, but if this primary fails then the new primary might have to
+                            // repeat phase 1 to recover this replica.
+                            // TODO TBD maybe do this earlier?
```
Should we do this earlier? We could do it before phase 1, with the advantage that it gives the other replicas a fighting chance to also retain the history needed to resume this recovery on a primary failure (still not guaranteed, mind) but the disadvantage of retaining extra history if this recovery fails and then we recover to a different node.

FWIW I prefer it here.

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -1556,14 +1556,18 @@ public long fileLength(String name) throws IOException {
-        assertThat(shard.refreshStats().getTotal(), equalTo(2L)); // refresh on: finalize and end of recovery
+        // refresh on: finalize and end of recovery
+        // finalizing a replica involves two refreshes with soft deletes because of estimateNumberOfHistoryOperations()
+        final long initialRefreshes = shard.routingEntry().primary() || shard.indexSettings().isSoftDeleteEnabled() == false ? 2L : 3L;
```
I'm not sure if this is expected or not. Seems a bit awkward. Maybe there's a better solution?

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -1590,18 +1594,18 @@ public void testExternalRefreshMetric() throws IOException {
-        long externalRefreshCount = shard.refreshStats().getExternalTotal();
-
+        final long externalRefreshCount = shard.refreshStats().getExternalTotal();
+        final long extraInternalRefreshes = shard.routingEntry().primary() || shard.indexSettings().isSoftDeleteEnabled() == false ? 0 : 1;
```
I'm not sure if this is expected or not. Seems a bit awkward. Maybe there's a better solution?

dnhatn commented on server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -158,13 +164,32 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
                     throw new DelayRecoveryException("source node does not have the shard listed in its state as allocated on the node");
                 }
                 assert targetShardRouting.initializing() : "expected recovery target to be initializing but was " + targetShardRouting;
+                retentionLeaseRef.set(useRetentionLeases ? shard.getRetentionLeases().get(
```
Is there any issue if we use any existing retention leases for the peer recovery purpose? I mean to rely on "Engine#getMinRetainedSeqNo".

DaveCTurner replied:
Yes, the issue is that we do not guarantee that the primary retains every operation required by every retention lease, so we must use `hasCompleteHistoryOperations` to ensure this. The problem is that in a file-based recovery we create a retention lease at the local checkpoint of this shard's safe commit, which may be behind every other lease, so we cannot be certain that every other peer is also able to respect this lease; if this primary were to fail then another primary may be elected without all the history needed for all its leases.

DaveCTurner replied:
See e.g. this comment: https://github.com/elastic/elasticsearch/pull/44853/files#diff-9feb4947ba4df0b486c7de6b9d46502fR215-R219

DaveCTurner replied:
On reflection, it might actually be possible to do this: we'd need to keep the retention lock open until at least having replayed history to the global checkpoint. I'll think about this a bit more.

DaveCTurner replied:
Ok, Yannick and I discussed this and you're right, there's no real drawback to creating the new replica's retention lease according to the current (persisted) global checkpoint rather than the local checkpoint of the safe commit. This means that we've much greater chance that this retention lease is satisfied on every in-sync shard copy, at which point we should not obviously need to call `hasCompleteHistoryOperations` in many cases.

It is, however, not totally watertight, because today we sometimes create unsatisfied leases for BWC reasons. I haven't gone through the details to see if there are other cases too, but given that it's not something we're asserting today I have concerns that it might be possible. Also it's cheap to call `hasCompleteHistoryOperations` and will save us from disaster, so let's keep it in.

I've adjusted the creation logic in 5fb8bda58b7ee62a8865ece9b1f1160142fd65c9.

dnhatn commented on server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -149,6 +151,10 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
             };
 
+            final boolean useRetentionLeases = shard.indexSettings().isSoftDeleteEnabled()
+                && shard.indexSettings().getIndexMetaData().getState() != IndexMetaData.State.CLOSE;
```
We also need to deal with frozen indices here.

DaveCTurner replied:
The reason we guard against closed indices here is because replicated write actions are blocked on closed indices, so we cannot create or remove any retention leases and must weaken this invariant in that case:

https://github.com/elastic/elasticsearch/blob/d7f7ebc7e2e7202e819e75edb88dbb56b662dee7/server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java#L750-L765

However this does not apply to frozen indices, and this invariant holds for them. This is despite the fact that we can't replay any operations into a frozen index, but that fact is true today in `master` as well. Could you clarify what extra changes you think we need here for frozen indices?

dnhatn replied:
okay, thanks for the explanation.

DaveCTurner commented on server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -158,13 +164,32 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
+                // temporarily prevent any history from being discarded, and do this before acquiring the safe commit so that we can
+                // be certain that all operations after the safe commit's local checkpoint will be retained for the duration of this
+                // recovery.
+                retentionLock = shard.acquireRetentionLock();
```
Oof good catch, thanks. Addressed in 190649d59bcf651b08b16f81833d7b140f8a8330

DaveCTurner commented on server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -255,6 +291,11 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
             }, onFailure);
 
             establishRetentionLeaseStep.whenComplete(r -> {
+                if (useRetentionLeases) {
```
It's not a no-op if we are doing a file-based recovery. In that case we have a proper retention lock at this point and should close it now so that we can start discarding history during phase 2. I clarified the condition to `useRetentionLeases && isSequenceNumberBasedRecovery == false` in 190649d59bc.

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java:
```diff
@@ -290,6 +290,35 @@ public RetentionLease addRetentionLease(
+        final RetentionLease retentionLease;
+        final RetentionLeases currentRetentionLeases;
+        synchronized (this) {
+            if (getRetentionLeases().contains(sourceLeaseId) == false) {
```
Thanks, yes, see bc5149a634d.

DaveCTurner commented on server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -158,13 +165,35 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
+            // without having a complete history.
+
+            if (isSequenceNumberBasedRecovery && useRetentionLeases) {
+                // all the history we need is retained by an existing retention lease, so we do not need a separate retention lock
```
No, leases for assigned shard copies never expire, so the lease will remain in place while the recovery is ongoing.

https://github.com/elastic/elasticsearch/blob/bc5149a634dc37ae7d3b9ec5aa43d4373aca5494/server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java#L238-L253

ywelsch replied:
🕺

DaveCTurner commented:
The test failure is meaningful: https://scans.gradle.com/s/bcfrvej2xrypy/console-log?task=:x-pack:plugin:ccr:internalClusterTest

What happened is that there was ongoing indexing while a replica was recovering, which plays out like this:

- we capture the GCP of the primary and use that in the `cleanFiles()` step when creating the new translog.

- the GCP of the primary advances, as does its retention lease.

- we clone the primary's (newly-advanced) retention lease for the replica.

- the replica's GCP does not advance.

The replica's retention lease is now ahead of its GCP, and this trips an assertion at renewal time. Yannick and I discussed a couple of options:

- weaken the assertion to ignore cases where the shard copy is not in sync. This makes it quite a bit weaker, but I can't see any terrible consequences of this.

- create the retention lease slightly earlier, within phase 1, and base the GCP of the replica's new translog on the retained seqno of this new retention lease.

@dnhatn WDYT?

dnhatn commented:
@DaveCTurner Thanks for the ping. Both options are good to me.

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/docs (failure looks unrelated and bogus)

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/1

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/1

DaveCTurner commented:
Ok I'm happy with this and it's passed a bunch of runs through `:server:test` and `:server:integtest` overnight. Worth a final pass.

Changes since the last reviews:

- https://github.com/elastic/elasticsearch/pull/44853/commits/d63e77774b37fd6bb20e5c5b2e2c3ffbbd1852ca to go back to today's behaviour of using the GCP of the primary as the starting GCP of the replica, but adjusting this to be sampled _after_ copying the files over and cloning the primary's lease so we can be sure that it's ahead of the leased checkpoint.

- https://github.com/elastic/elasticsearch/pull/44853/commits/f47e56eaf8f2cf69ad8008f7cd0bc1dcd7437897 to also create the lease when phase 1 is a no-op thanks to a synced flush marker (caught by BWC tests, but I added a proper test for it too).
